### ISSUE TITLE
Add support for pre-step and post-step logic

### ISSUE THREAD

**Issue by isaac-mason** [pmndrs/react-three-rapier#177]
Hi!

Looking for thoughts on adding pre-step and post-step events.

Use cases for this include creating custom character and vehicle controllers, where forces should only be applied once per physics step. When not using `timestep="vary"`, applying forces on each frame isn't desirable, as not every frame will step the physics world, some will just interpolate.

This could come in the form of hooks, e.g.
```ts
import { usePreStep, usePostStep } from "@react-three/rapier"

usePreStep(() => { /* called before stepping the world */ })
usePostStep(() => { /* called after stepping the world */ })
```

Or via signals/event handlers on the wrapped rapier `World`, e.g.
```ts
world.onPreStep.add(() => { /* called before stepping the world */ })
world.onPostStep.add(() => { /* called after stepping the world */ })
```

Or maybe a combination of both.

Keen to hear thoughts, and as usual, happy to help implement 🙂

**Comment by hmans** [comment]
Should this not be covered by the new fixed step stages in v9 (cc @krispya)? I'm not sure how far off v9 is, so it might still be useful to have a library-side implementation of this or something like it.

Also, this library recently added [manual stepping](https://github.com/pmndrs/react-three-rapier/#manual-stepping). This would allow the user to (even with v8) implement their own fixed stepping with code running before and after.

(Not necessarily an argument for or against your proposed feature; but please do take the potential overlap into consideration.)

**Comment by isaac-mason** [comment]
Had a closer look. It's currently possible to use `timeStep="vary"` to address this, with or without the manual step function.

Like you said, the manual step feature can be used, and code can be placed before and after. An `updatePriority` can also be provided to the main `Physics` component, so another `useFrame` could be used with a priority higher or lower than the `Physics` `updatePriority`.

The issue comes in when interpolation is enabled. When interpolating, each update may not result in a physics step. As far as I can see, there's nothing currently exposed that could tell me whether for a given update, a step has occurred, or whether rt/rapier has just interpolated.

**Comment by krispya** [comment]
Could the interpolation be handled on the JS side? When integrated with v9 stages I think this would give the most flexibility for devs.

**Comment by isaac-mason** [comment]
From discussions on discord with @krispya (let me know if I've misrepresented anything 🙂):

---

A rt/fiber v9 friendly way of handling this could be as follows:
- Split physics steps and physics body interpolation into separate operations
- Use v9 stages for physics steps and interpolation
- Support overriding of default stages

This could enable lifecycle configuration like this: 

```ts
import { Stage, FixedStage } from "@react-three/fiber"

const physicsTimeStep = 1 / 50

// create a pre-step stage that should run only once before each physics step
const PhysicsPreStepStage = new FixedStage(physicsTimeStep)

// create a physics step stage
// this will be provided to rt/rapier as an override for the default physics step stage
// could be `FixedStage` for fixed stepping, and `Stage` for `timeStep="vary"` equivalent
const PhysicsStepStage = new FixedStage(physicsTimeStep)

// explictly create a stage for interpolation
// this will be provided to rt/rapier as an override for the default interpolation stage
// could be omitted, but included here to highlight that physics steps and interpolation are separate in this example
const PhysicsInterpolationStage = new Stage() 

const lifecycle = [
  // ...
  PhysicsPreStepStage, // place pre-step stage before physics step stage
  PhysicsStepStage,
  PhysicsInterpolationStage,
  // ...
]

const App = () => (
  <Canvas stages={lifecycle}>
    <Physics stages={{
      // provide the stage overrides to rt/rapier
      step: PhysicsStepStage,
      interpolation: PhysicsInterpolationStage // 'interpolation' could be omitted, defaulting to 'Stage.Update'
    }}>
      { /* ... */ }
    </Physics>
  </Canvas>
)
```

Then, adding pre-step logic look like this:

```ts
useUpdate(() => {
  /* your pre-step logic here */
}, PhysicsPreStepStage)
```

**Comment by wiledal** [comment]
I think the solution above ☝️  works, but at the same time it's a bit overly complicated, whilst not being backwards compatible.

If this is something we want _now_ I think exposing some hooks would be simple. A v9 upgrade could use a `FixedStage` further down.

I like the idea of having simple hooks exposed, like you initially suggested @isaac-mason.

`useBeforePhysicsStep(() => {...})`
`useAfterPhysicsStep(() => {...})`

**Comment by krispya** [comment]
Yeah, it's an interesting problem space. I added the ability to customize the stage lifecycle but it's really something for a power user. I don't think people should have to think about it when interacting with a library. My suggestion for both mental model and backwards compatibility is to make something like those hooks as suggested and then with v9 have the ability to split out those loops into their own stages if the user is so inclined. Otherwise they get slotted into lifecycle automatically be default. 

I think the part that isn't clear from just the hooks is that `useBeforePhysicsStep` would presumably be inside the same loop as the physics step, meaning it can be fixed or unconstrained while `useAfterPhysicsStep`, if it is meant for interpolation, has to always be in an unconstrained loop. This would be a big gotcha for people who aren't intimate with the Rapier lifecycle as `useAfterPhysicsStep` would not be an actual effect.

Maybe a way to configure the loops these physics stages fall under (fixed or "varying) would make it more clear and also preempt more robust customization with v9.

**Comment by isaac-mason** [comment]
I think both `useBeforePhysicsStep` and `useAfterPhysicsStep` should be within the same loop as the physics step - running only once for a rapier physics step.

To get interpolated values, you could already use `useFrame` with a lower priority than the `<Physics />` `updatePriority`.

**Comment by isaac-mason** [comment]
I'll start with creating a PR adding `useBeforePhysicsStep` and `useAfterPhysicsStep` shortly then.

**Comment by krispya** [comment]
> I think both `useBeforePhysicsStep` and `useAfterPhysicsStep` should be within the same loop as the physics step - running only once for a rapier physics step.
> 
> To get interpolated values, you could already use `useFrame` with a lower priority than the `<Physics />` `updatePriority`.
> 
> (Correct me if you don't think this is right)

Makes perfect sense to me.

**Comment by isaac-mason** [comment]
Little update - I'm still planning to make a PR for this, but my first use case will be the raycast vehicle controller I've been working on. Once that's done, I'll circle back to this as part of building a rt/rapier raycast vehicle controller addon.

**Comment by wiledal** [comment]
I've added a draft for this in #258 @isaac-mason, as preparation for the `addons` package, also needed for the `Attractor` separation. What do you think?
